doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,401 @@
+
+    Note that Threshold can be configured to a very large number so that only the scenario (1) occurs.
+    With this design, we can control the overall I/O procedure with the optimizations for crimson
+    as described above.
+
+  - Detailed flow
+    - 1. Append a log entry that contains pg_log, snapset, object_infot_t, block allocation
+      using NVMe atomic write command on the WAL
+      - NVMe provides atomicity guarantees for a write command (Atomic Write Unit Power Fail)
+        For example, 512 Kbytes of data can be atomically written at once without fsync()
+      - Small size (object_info_t, snapset, collection, etc.) can be embed
+      - stage 1
+        WAL (written) --> | TxBegin A | Log Entry | TxEnd A | (if small)
+        Data partition (written)--> | Data blocks | (if large)
+    - 2. Then, append the data to WAL (if large)
+      - stage 2 (Updating object meta and allocation bitmap can be skipped via pre-allocation)
+        WAL --> | TxBegin A | Log Entry | TxEnd A | (if large)
+
+- Read
+  - Use the cached object metadata to find out the data location
+  - If not cached, need to search WAL after checkpoint and Object meta partition to find the 
+    latest meta data
+
+- Flush
+  - Flush WAL entries whenever needed
+
+
+Crash consistency
+------------------
+Large case
+- 1. Crash occurs right after writing Data blocks
+  - Data partition --> | Data blocks |
+  - We don't need to care this case. Data is not alloacted yet in reality. The blocks will be reused.
+- 2. Crash occurs right after WAL using atomic write command
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - Data partition --> | Data blocks |
+  - Write procedure is completed, so there is no data loss or inconsistent state
+
+Small case
+- 1. Crash occurs right after writing WAL
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - All data has been written
+
+
+Comparison
+----------
+- Best case (pre-allocation)
+  - Only need two writes on both WAL and Data partition without updating Object meta and Data bitmap
+- Worst case (flush happens)
+  - Without embedding, at least three writes are required additionally on Object meta, Data bitmap, and Data blocks
+- WAL needs to be flushed if the WAL is close to full and the WAL entry is snapset, object_info_t, 
+  collection and pg_log
+  - The premise behind this design is OSD can manage the lateset metadta as a single copy. So,
+    appended entires are not to be read
+- Either best of worst case does not produce severe I/O amplification (it produce I/Os, but I/O rate is constant) 
+  unlike LSM-tree DB (proposed design is similar to LSM-tree which has only level-0)
+  
+
+Detailed Design 
+===============
+
+- Onode lookup
+ - Flat namespace
+  Object data, object meta, omap, and xattrs are all the data related to an object (we call it a head object from ghobject).
+  PoseidonStore adds different suffixes to the head object ID to create PSD_OID to distinguish them
```

Comment:
So your metadata btree stores arbitrary length strings sorted lexicographically?

Reply:
How do you plan on addressing COW clones?